src/v/iceberg/compatibility.cc

```diff
@@ -502,4 +506,184 @@ try_fill_field_ids(const struct_type& source, struct_type& dest) {
+                    // Writer type is promotable to/compatible with the host
+                    // type.
+                    return outcome::success();
+                case type_promoted::changes_partition:
+                    // TODO: Update the return type so that this impossible
+                    // value does not need to be handled.
+                    return schema_evolution_errc::incompatible;
+                }
+            } else {
+                // writer -> host promotion fails, let's check the opposite
+                // direction and upgrade/promote the field if needed.
+                auto res = std::visit(
+                  primitive_type_promotion_policy_visitor_v,
+                  host_type,
+                  writer_type);
+                if (res.has_error()) {
+                    // Errors in both directions. The fields are incompatible at
+                    // all/unrelated types.
+                    return schema_evolution_errc::incompatible;
+                } else {
+                    switch (res.value()) {
+                    case type_promoted::no:
+                        // Both types are the same.
+                        return outcome::success();
+                    case type_promoted::yes:
+                        host_type = writer_type; // promote the type
+                        return outcome::success();
+                    case type_promoted::changes_partition:
+                        // TODO: Update the return type so that this impossible
+                        // value does not need to be handled.
+                        return schema_evolution_errc::incompatible;
+                    }
+                }
+            }
+        }
+    }
+
+    schema_merge_result operator()(
+      const list_type& writer_list_type,
+      const list_type& host_list_type) const {
+        const auto& writer_element = writer_list_type.element_field;
+        auto& host_element = host_list_type.element_field;
+
+        // Merge list element type recursively.
+        vassert(
+          writer_element != nullptr && host_element != nullptr,
+          "List element fields assumed to be non-NULL");
+        if (auto ve_res = std::visit(
+              *this, writer_element->type, host_element->type);
+            ve_res.has_error()) {
+            return ve_res.error();
+        }
+        return outcome::success();
+    }
+
+    schema_merge_result operator()(
+      const map_type& writer_map_type, const map_type& host_map_type) const {
+        const auto& writer_key = writer_map_type.key_field;
+        auto& host_key = host_map_type.key_field;
+        const auto& writer_value = writer_map_type.value_field;
+        auto& host_value = host_map_type.value_field;
+
+        vassert(
+          writer_key != nullptr && host_key != nullptr,
+          "Map key fields assumed to be non-NULL");
+        vassert(
+          writer_value != nullptr && host_value != nullptr,
```

Comment:
Why don't we need to do promotion on the value fields? Mind leaving a comment explaining, if that's intentional?